code/datums/weakref.dm

```diff
@@ -0,0 +1,33 @@
+/datum
+	var/weakref/weakref
+
+/datum/Destroy()
+	weakref = null // Clear this reference to ensure it's kept for as brief duration as possible.
+	. = ..()
+
+//obtain a weak reference to a datum
+/proc/weakref(datum/D)
+	if(!istype(D))
+		return
+	if(QDELETED(D))
+		return
+	if(!D.weakref)
+		D.weakref = new/weakref(D)
+	return D.weakref
+
+/weakref
+	var/ref
+
+/weakref/New(datum/D)
+	ref = "\ref[D]"
```

Comment:
`ref = "[REF(D)]"`

In order to be compatible with 512 all uses of \ref need to be REF() instead

Reply:
Why `"[]"`? Doesn't `REF()` output strings already?

Reply:
well yes but I was generalizing for the other \refs that'll need to be replaced